text/0105-xcm-improved-fee-mechanism.md

```diff
@@ -0,0 +1,112 @@
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 23 July 2024                                                                                |
+| **Description** | Allow multiple types of fees to be paid                                                     |
+| **Authors**     | Francisco Aguirre                                                                           |
+
+## Summary
+
+XCM already handles execution fees in an effective and efficient manner using the `BuyExecution` instruction.
+However, other types of fees are not handled as effectively -- for example, delivery fees.
+Fees exist that can't be measured using `Weight` -- as execution fees can -- so a new method should be thought up for those cases.
+This RFC proposes making the fee handling system simpler and more general, by doing two things:
+- Adding a `fees` register
+- Deprecating `BuyExecution` and adding a new instruction `PayFees` with new semantics to ultimately replace it.
+
+## Motivation
+
+Execution fees are handled correctly by XCM right now.
+However, the addition of extra fees, like for message delivery, result in awkward ways of integrating them into the XCVM implementation.
+This is because these types of fees are not included in the language.
+The standard should have a way to correctly deal with these implementation specific fees, that might not exist in every system that uses XCM.
+The new instruction moves the specified amount of fees from the holding register to a dedicated fees register that the XCVM can use in flexible ways depending on its implementation.
+The XCVM implementation is free to use these fees to pay for execution fees, transport fees, or any other type of fee that might be necessary.
+This moves the specifics of fees further away from the XCM standard, and more into the actual underlying XCVM implementation, which is a good thing.
+
+## Stakeholders
+
+- Runtime Users
+- Runtime Devs
+- Wallets
+- dApps
+
+## Explanation
+
+The new instruction that will replace `BuyExecution` is a much simpler and general version: `PayFees`.
+This instruction takes one `Asset`, takes it from the holding register, and puts it into a new `fees` register.
+The XCVM implementation can now use this `Asset` to make sure every necessary fee is paid for, this includes execution fees, delivery fees, or any other fee.
+
+```rust
+PayFees { asset: Asset }
+```
+
+This new instruction will reserve **the entirety** of `asset` for fee payment.
+The asset can't be used for anything else during the entirety of the program.
+This is different from the current semantics of `BuyExecution`.
+
+If not all `Asset` in the `fees` register is used when the execution ends, then we trap them alongside any possible leftover assets from the holding register.
```

Comment:
do we need to mention that `RefundSurplus` can be used to move asset from the fees register to holding register? and programs are expected to use RefundSurplus to avoid trapped assets